src/charm.py

```diff
@@ -1,187 +1,201 @@
-                            'resources': ['persistentvolumeclaims'],
-                            'verbs': ['create', 'delete'],
-                        },
-                    ],
-                }]
-            },
-            # "restartPolicy": 'Always',
-            # "terminationGracePeriodSeconds": 10,
-        })
-        self.model.unit.status = ActiveStatus()
-        return
-
-    def sanitized_container_config(self):
-        """Uninterpolated container config without secrets"""
-        config = self.framework.model.config
-        if config["container_config"].strip() == "":
-            container_config = {}
-        else:
-            container_config = \
-                yaml.safe_load(self.framework.model.config["container_config"])
-            if not isinstance(container_config, dict):
-                self.framework.model.unit.status = \
-                    BlockedStatus("container_config is not a YAML mapping")
-                return None
-        return container_config
-
-
-def log(message, level=None):
-    """Write a message to the juju log"""
-    command = ['juju-log']
-    if level:
-        command += ['-l', level]
-    if not isinstance(message, str):
-        message = repr(message)
-
-    # https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/binfmts.h
-    # PAGE_SIZE * 32 = 4096 * 32
-    MAX_ARG_STRLEN = 131072
-    command += [message[:MAX_ARG_STRLEN]]
-    # Missing juju-log should not cause failures in unit tests
-    # Send log output to stderr
-    subprocess.call(command)
-
-
-if __name__ == '__main__':
+            }
+        }
+
+    def _configure_pod(self):
+        """Setup a new Microsoft SQL Server pod specification"""
+        if not self.unit.is_leader():
+            self.unit.status = self.UNIT_ACTIVE_STATUS
+            return
+
+        if not self._validate_config():
+            logger.warning('Charm config is not valid')
+            return
+
+        logger.info("Setting pod spec")
+        self.unit.status = MaintenanceStatus('Setting pod spec')
+
+        pod_spec = self._build_pod_spec()
+        pod_resources = self._build_pod_resources()
+        self.model.pod.set_spec(pod_spec, pod_resources)
+
+        self.app.status = self.APPLICATION_ACTIVE_STATUS
+        self.unit.status = self.UNIT_ACTIVE_STATUS
+
+    def _sa_password(self, length=32):
+        sa_pass = leader_get('sa_password')
+        if sa_pass:
+            return sa_pass
+
+        random_len = math.ceil(length/4)
+        lower = ''.join(random.choice(string.ascii_lowercase)
+                        for i in range(random_len))
+        upper = ''.join(random.choice(string.ascii_uppercase)
+                        for i in range(random_len))
+        digits = ''.join(random.choice(string.digits)
+                         for i in range(random_len))
+        special = ''.join(random.choice(string.punctuation)
+                          for i in range(random_len))
```

Comment:
you might want to consider using `secrets.choice` instead of `random.choice`, if python 3.5 support isn't needed for the charm

Reply:
Sounds good! The official MSSQL Docker image runs Ubuntu 18.04, so I think we're fine with `secrets.choice`.